Clean/Circomlib/Gates.lean

```diff
@@ -259,4 +270,781 @@ def circuit : FormalCircuit (F p) fieldPair field where
+    | 0 =>
+      intros
+      simp [main, Circuit.operations, Circuit.pure_def]
+      constructor <;> intro <;> trivial
+    | 1 =>
+      intros
+      simp [main, Circuit.operations, Circuit.pure_def]
+      constructor <;> intro <;> trivial
+    | 2 =>
+      intros
+      subst offset2
+      simp only [main]
+      constructor
+      ·
+        intro h_witnesses
+        apply Environment.can_replace_usesLocalWitnessesCompleteness
+        · apply AND.circuit.subcircuitsConsistent
+        · exact h_witnesses
+      · intro h_completeness
+        simp only [AND.circuit, subcircuit_norm, AND.main, bind_pure, Fin.isValue, bind_pure_comp, circuit_norm] at h_completeness ⊢
+        simp only [Fin.isValue, Nat.add_zero, id_eq]
+        unfold Environment.UsesLocalWitnesses Operations.forAllFlat
+        unfold Operations.forAll
+
+
+        constructor
+        · simp only [Environment.ExtendsVector, Vector.getElem_mk]
+          intro i
+          fin_cases i
+          simp only [Fin.val_zero, add_zero, List.getElem_toArray]
+          exact h_completeness
+        · simp only [Operations.forAll]
+          trivial
+    | m + 3 =>
+      intros
+      subst offset2
+      simp only [main]
+      constructor
+      · intro h_witnesses
+        let n1 := (m + 3) / 2
+        let n2 := (m + 3) - n1
+        have h_n1_lt : n1 < m + 3 := Nat.div_lt_self (by omega) (by norm_num)
+        have h_n2_lt : n2 < m + 3 := by simp only [n2, n1]; omega
+        apply Environment.can_replace_usesLocalWitnessesCompleteness
+        · rw [← main]
+          apply subcircuitsConsistent
+        · exact h_witnesses
+      · intro h_completeness
+        simp only [circuit_norm, subcircuit_norm] at h_completeness ⊢
+        rcases h_completeness with ⟨ h_c1, h_c2, h_c3 ⟩
+
+        rw[Environment.UsesLocalWitnesses, Operations.forAllFlat]
+
+        rw [Operations.forAll_append]
+        constructor
+        · rw[← Operations.forAllFlat, ← Environment.UsesLocalWitnesses]
+          rw[IH]
+          · aesop
+          · omega
+          · trivial
+        rw [Operations.forAll_append]
+        constructor
+        · rw[← Operations.forAllFlat, ← Environment.UsesLocalWitnesses]
+          rw[IH]
+          · aesop
+          · omega
+          · omega
+        · simp only [AND.circuit] at h_c3 ⊢
+          simp only [AND.main, circuit_norm] at h_c3 ⊢
+          constructor
+          · exact h_c3
+          · simp only [Gadgets.Equality.circuit, Gadgets.Equality.elaborated, Gadgets.Equality.main, subcircuit_norm]
+            rw [Circuit.forEach]
+            simp_all [toVars, assertZero, var, circuit_norm, Operations.toFlat, FlatOperation.forAll]
+
+-- Extract Assumptions and Spec outside the circuit
+def Assumptions (n : ℕ) (input : fields n (F p)) : Prop :=
+  ∀ (i : ℕ) (h : i < n), IsBool input[i]
+
+def Spec (n : ℕ) (input : fields n (F p)) (output : F p) : Prop :=
+  output.val = (input.toList.map (·.val)).foldl (· &&& ·) 1 ∧ IsBool output
```

Comment:
I think `toList` here is causing lots of unnecessary `Vector.toList` conversions.